File: backend/src/state_computation.c
/*
 * state_computation.c — Phase 2: COMPUTE_OPTIMAL_STRATEGY
 *
 * Backward induction orchestrator. Processes states from |C|=15 (terminal)
 * down to |C|=0 (game start), calling SOLVE_WIDGET for each state.
 *
 * See: theory/optimal_yahtzee_pseudocode.md, Phase 2: COMPUTE_OPTIMAL_STRATEGY.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include "state_computation.h"
#include "storage.h"
#include "phase0_tables.h"
#include "widget_solver.h"
#include "timing.h"

/* ── Progress tracking ───────────────────────────────────────────────── */

typedef struct {
    int total_states;
    int completed_states;
    double start_time;
    double last_report_time;
    int states_per_level[16];
    double time_per_level[16];
} ComputeProgress;

static void InitProgress(ComputeProgress *progress) {
    memset(progress, 0, sizeof(ComputeProgress));
    progress->start_time = timer_now();
    progress->last_report_time = progress->start_time;

    for (int num_scored = 0; num_scored <= 15; num_scored++) {
        int states = 0;
        for (int scored = 0; scored < (1 << CATEGORY_COUNT); scored++) {
            if (__builtin_popcount(scored) == num_scored) {
                states += 64;
            }
        }
        progress->states_per_level[num_scored] = states;
        progress->total_states += states;
    }
}

static void UpdateProgress(ComputeProgress *progress, int level, int states_completed) {
    progress->completed_states += states_completed;
    progress->time_per_level[level] = timer_now() - progress->start_time;
}

static void PrintProgress(ComputeProgress *progress, int current_level) {
    double now = timer_now();
    if (progress->completed_states < progress->total_states &&
        now - progress->last_report_time < 0.5) {
        return;
    }
    progress->last_report_time = now;

    double elapsed = now - progress->start_time;
    double pct = (double)progress->completed_states / progress->total_states * 100.0;
    double rate = progress->completed_states / elapsed;
    double eta = (progress->total_states - progress->completed_states) / rate;

    printf("\rProgress: %d/%d states (%.1f%%) | Level: %d | Elapsed: %.1fs | Rate: %.0f states/s | ETA: %.1fs     ",
           progress->completed_states, progress->total_states, pct,
           current_level, elapsed, rate, eta);
    fflush(stdout);
}

/* ── Phase 2 main loop ───────────────────────────────────────────────── */

void ComputeAllStateValues(YatzyContext *ctx) {
    ComputeProgress progress;
    InitProgress(&progress);

    printf("=== Starting State Value Computation ===\n");
    printf("Total states to compute: %d\n", progress.total_states);
    printf("States per level:\n");
    for (int i = 0; i <= 15; i++) {
        printf("  Level %2d: %6d states\n", i, progress.states_per_level[i]);
    }
    printf("\n");

    double total_start = timer_now();

    /* Try to load consolidated file first */
    const char *consolidated_file = "data/all_states.bin";
    if (LoadAllStateValuesMmap(ctx, consolidated_file)) {
        printf("Loaded pre-computed states from consolidated file\n");
        return;
    }

    /* Process states level by level, from game end (15) to game start (0) */
    for (int num_scored = 15; num_scored >= 0; num_scored--) {
        double level_start = timer_now();
        char level_filename[64];
        snprintf(level_filename, sizeof(level_filename), "data/states_%d.bin", num_scored);

        PrintProgress(&progress, num_scored);

        /* Level 15: terminal states (base case) */
        if (num_scored == 15) {
            if (!LoadStateValuesForCount(ctx, num_scored, level_filename)) {
                InitializeFinalStates(ctx);
                SaveStateValuesForCount(ctx, num_scored, level_filename);
            }
            UpdateProgress(&progress, num_scored, progress.states_per_level[num_scored]);
            continue;
        }

        /* Try loading from disk */
        if (LoadStateValuesForCount(ctx, num_scored, level_filename)) {
            UpdateProgress(&progress, num_scored, progress.states_per_level[num_scored]);
            continue;
        }

        /* Compute this level */
        printf("\nComputing level %d (%d states)...\n",
               num_scored, progress.states_per_level[num_scored]);

        int states_to_compute = progress.states_per_level[num_scored];
        int (*state_list)[2] = malloc(sizeof(int) * 2 * states_to_compute);
        if (!state_list) {
            fprintf(stderr, "Memory allocation failed for level %d\n", num_scored);
            return;
        }

        int state_index = 0;
        for (int scored = 0; scored < (1 << CATEGORY_COUNT); scored++) {
            if (__builtin_popcount(scored) == num_scored) {
                for (int up = 0; up <= 63; up++) {
                    state_list[state_index][0] = up;
                    state_list[state_index][1] = scored;
                    state_index++;
                }
            }
        }

        const int BATCH = 1000;

        #pragma omp parallel for schedule(dynamic, 64)
        for (int i = 0; i < state_index; i++) {
            int up = state_list[i][0];
            int scored = state_list[i][1];

            YatzyState current = {up, scored};
            double ev = ComputeExpectedStateValue(ctx, &current);
            ctx->state_values[STATE_INDEX(up, scored)] = ev;

            if (i % BATCH == 0) {
                #pragma omp critical
                {
                    UpdateProgress(&progress, num_scored, BATCH);
                    PrintProgress(&progress, num_scored);
                }
            }
        }

        int remaining = state_index % BATCH;
        if (remaining > 0) {
            UpdateProgress(&progress, num_scored, remaining);
        }

        free(state_list);
        SaveStateValuesForCount(ctx, num_scored, level_filename);

        double level_dur = timer_now() - level_start;
        printf("\nLevel %d completed in %.2f seconds (%.0f states/sec)\n",
               num_scored, level_dur, progress.states_per_level[num_scored] / level_dur);
    }

    printf("\n\n=== Computation Complete ===\n");

    printf("\nSaving consolidated state file...\n");
    SaveAllStateValuesMmap(ctx, consolidated_file);

    double total_time = timer_now() - total_start;
    printf("\nTotal computation time: %.2f seconds\n", total_time);
    printf("Average processing rate: %.0f states/second\n", progress.total_states / total_time);

    printf("\nDetailed timing breakdown by level:\n");
    printf("Level | States  | Time (s) | Rate (states/s)\n");
    printf("------|---------|----------|----------------\n");

    double prev_time = 0;
    for (int level = 15; level >= 0; level--) {
        double level_time = progress.time_per_level[level] - prev_time;
        if (level_time > 0) {
            printf("  %2d  | %6d  | %7.2f  | %6.0f\n",
                   level, progress.states_per_level[level], level_time,
                   progress.states_per_level[level] / level_time);
        }
        prev_time = progress.time_per_level[level];
    }
}

File: backend/src/state_computation.rs
//! Phase 2: Backward induction — compute E_table[S] for all reachable states.
//!
//! Implements pseudocode `COMPUTE_OPTIMAL_STRATEGY`. Processes states in decreasing
//! order of |C| (number of scored categories), from |C|=14 down to |C|=0. Terminal
//! states (|C|=15) are initialized in Phase 0.
//!
//! Each level is parallelized with rayon `par_iter`. Within each state, SOLVE_WIDGET
//! reads only from successor states (|C|+1), which are already computed — the same
//! guarantee that enables parallelism in the pseudocode.
//!
//! ## Unsafe writes
//!
//! Each (upper_score, scored_categories) maps to a unique `state_index`, so parallel
//! workers never write to the same memory location. We use `AtomicPtr` + unsafe raw
//! pointer writes to avoid the overhead
//! of `collect()` + scatter.

use std::time::Instant;

use rayon::prelude::*;

use crate::constants::*;
use crate::storage::{load_all_state_values, save_all_state_values};
use crate::types::{YatzyContext, YatzyState};
use crate::widget_solver::compute_expected_state_value;

/// Progress tracker for the Phase 2 computation loop.
struct ComputeProgress {
    total_states: i32,
    completed_states: i32,
    start_time: Instant,
    last_report_time: Instant,
    states_per_level: [i32; 16],
    time_per_level: [f64; 16],
}

impl ComputeProgress {
    fn new(ctx: &YatzyContext) -> Self {
        let now = Instant::now();
        let mut progress = ComputeProgress {
            total_states: 0,
            completed_states: 0,
            start_time: now,
            last_report_time: now,
            states_per_level: [0; 16],
            time_per_level: [0.0; 16],
        };

        for num_scored in 0..=15 {
            let mut states = 0;
            for scored in 0..(1u32 << CATEGORY_COUNT) {
                if scored.count_ones() == num_scored as u32 {
                    let upper_mask = (scored & 0x3F) as usize;
                    for up in 0..=63 {
                        if ctx.reachable[upper_mask][up] {
                            states += 1;
                        }
                    }
                }
            }
            progress.states_per_level[num_scored] = states;
            progress.total_states += states;
        }

        progress
    }

    fn print_progress(&mut self, current_level: i32) {
        let now = Instant::now();
        if self.completed_states < self.total_states
            && now.duration_since(self.last_report_time).as_secs_f64() < 0.5
        {
            return;
        }
        self.last_report_time = now;

        let elapsed = now.duration_since(self.start_time).as_secs_f64();
        let pct = self.completed_states as f64 / self.total_states as f64 * 100.0;
        let rate = self.completed_states as f64 / elapsed;
        let eta = (self.total_states - self.completed_states) as f64 / rate;

        print!(
            "\rProgress: {}/{} states ({:.1}%) | Level: {} | Elapsed: {:.1}s | Rate: {:.0} states/s | ETA: {:.1}s     ",
            self.completed_states, self.total_states, pct, current_level, elapsed, rate, eta
        );
        use std::io::Write;
        let _ = std::io::stdout().flush();
    }
}

/// Compute E_table[S] for all reachable game states using backward induction.
///
/// Pseudocode: `COMPUTE_OPTIMAL_STRATEGY` — iterates num_filled from 14 down to 0,
/// calling SOLVE_WIDGET for each valid state at that level. Level 15 (terminal) is
/// already initialized.
///
/// Results are saved to `data/all_states.bin`.
pub fn compute_all_state_values(ctx: &mut YatzyContext) {
    let mut progress = ComputeProgress::new(ctx);

    println!("=== Starting State Value Computation ===");
    println!(
        "Total states to compute: {} (after reachability pruning)",
        progress.total_states
    );
    println!("States per level:");
    for i in 0..=15 {
        println!("  Level {:2}: {:6} states", i, progress.states_per_level[i]);
    }
    println!();

    let total_start = Instant::now();

    // Try to load consolidated file first
    let consolidated_file = "data/all_states.bin";
    if load_all_state_values(ctx, consolidated_file) {
        println!("Loaded pre-computed states from consolidated file");
        return;
    }

    // Process states level by level, from game end (14) to game start (0).
    // Level 15 (terminal) is already initialized by precompute_lookup_tables.
    for num_scored in (0..=14).rev() {
        let level_start = Instant::now();

        progress.print_progress(num_scored);

        println!(
            "\nComputing level {} ({} states)...",
            num_scored, progress.states_per_level[num_scored as usize]
        );

        // Build state list grouped by scored_categories mask.
        // States sharing a scored mask access the same 256-byte successor regions
        // (due to scored*64+up index layout), so grouping them maximizes L1 cache hits.
        let mut groups: Vec<(i32, Vec<usize>)> = Vec::new();
        for scored in 0..(1u32 << CATEGORY_COUNT) {
            if scored.count_ones() == num_scored as u32 {
                let upper_mask = (scored & 0x3F) as usize;
                let ups: Vec<usize> = (0..=63usize)
                    .filter(|&up| ctx.reachable[upper_mask][up])
                    .collect();
                if !ups.is_empty() {
                    groups.push((scored as i32, ups));
                }
            }
        }

        // Parallel computation: each group processes sequentially on one thread.
        // Safety: each (up, scored) pair maps to a unique state_index,
        // so no two threads write to the same location.
        let sv_ptr = ctx.state_values.as_mut_slice().as_mut_ptr();

        // AtomicPtr wrapper to satisfy Send+Sync for rayon's for_each.
        // We only use load/store — no actual atomic contention since indices are disjoint.
        let sv_atomic = std::sync::atomic::AtomicPtr::new(sv_ptr);

        let ctx_ref = &*ctx;
        groups.par_iter().for_each(|(scored, ups)| {
            let ptr = sv_atomic.load(std::sync::atomic::Ordering::Relaxed);
            for &up in ups {
                let state = YatzyState {
                    upper_score: up as i32,
                    scored_categories: *scored,
                };
                let ev = compute_expected_state_value(ctx_ref, &state);
                unsafe {
                    *ptr.add(state_index(up, *scored as usize)) = ev as f32;
                }
            }
        });

        let state_count: usize = groups.iter().map(|(_, ups)| ups.len()).sum();

        let pre_loop_completed = progress.completed_states;
        progress.completed_states = pre_loop_completed + state_count as i32;
        progress.time_per_level[num_scored as usize] = total_start.elapsed().as_secs_f64();

        let level_dur = level_start.elapsed().as_secs_f64();
        println!(
            "\nLevel {} completed in {:.2} seconds ({:.0} states/sec)",
            num_scored,
            level_dur,
            progress.states_per_level[num_scored as usize] as f64 / level_dur
        );

        #[cfg(feature = "timing")]
        {
            crate::widget_solver::print_timing();
            crate::widget_solver::reset_timing();
        }
    }

    println!("\n\n=== Computation Complete ===");

    println!("\nSaving consolidated state file...");
    save_all_state_values(ctx, consolidated_file);

    let total_time = total_start.elapsed().as_secs_f64();
    println!("\nTotal computation time: {:.2} seconds", total_time);
    println!(
        "Average processing rate: {:.0} states/second",
        progress.total_states as f64 / total_time
    );

    println!("\nDetailed timing breakdown by level:");
    println!("Level | States  | Time (s) | Rate (states/s)");
    println!("------|---------|----------|----------------");

    let mut prev_time = 0.0;
    for level in (0..=15).rev() {
        let level_time = progress.time_per_level[level] - prev_time;
        if level_time > 0.0 {
            println!(
                "  {:2}  | {:6}  | {:7.2}  | {:6.0}",
                level,
                progress.states_per_level[level],
                level_time,
                progress.states_per_level[level] as f64 / level_time
            );
        }
        prev_time = progress.time_per_level[level];
    }
}
